### PR TITLE
Add support for IPython/Jupyter notebook magics

### PR DESCRIPTION
**Pull request by mlucool**
Example usage (works in ipython/notebook/lab):
```
%load_ext pyinstrument
```

```
%%pyinstrument
def a():
    b()
    c()
def b():
    d()
def c():
    d()
def d():
    e()
def e():
    time.sleep(1)
a()
```
image

Extra care is taken to not profile jupyter running the code, which
is the majority of the complexity of this commit.

Fixes: #153

### CODE REVIEW

**Comment by joerick** [comment]
I'd like to have a go at this, seems like there might be a way to incorporate a test based on your starting point.

**Comment by joerick** [comment]
So. I have the IPython test working, but for some reason, even importing the IPython codebase on Windows causes a memory/stack overflow crash in the overflow test. No clue what's going on there, but it's probably not worth the time to debug. IPython is more of an application than a library, I expect there's some global/system manipulation going on that's causing an issue. I'll just sidestep the whole thing and run the IPython tests in their own process.

File: test/test_overflow.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import faulthandler
```
Comment (joerick) [inline test/test_overflow.py:1]:
```suggestion
```

**Comment by joerick** [comment]
Cool, I think this is ready. Does it look good to you @mlucool?

**Comment by mlucool** [comment]
This looks great! Thanks for all the work on this!

**Comment by mlucool** [comment]
Thanks for merging @joerick. When do you expect to release this?

**Comment by joerick** [comment]
Should be in the next few days, I want to get #160 in and then we should be good to go

**Pull request merged**